src/server/ua_server.c

```diff
@@ -521,65 +688,201 @@ UA_Server_removeCallback(UA_Server *server, UA_UInt64 callbackId) {
 }
 
 static void
-notifySecureChannelsStopped(struct UA_ServerComponent *sc,
-                            UA_LifecycleState state) {
-    if(sc->state == UA_LIFECYCLESTATE_STOPPED &&
-       sc->server->state == UA_LIFECYCLESTATE_STARTED) {
-        sc->notifyState = NULL; /* remove the callback */
-        sc->start(sc, sc->server);
+secureChannel_delayedClose(void *application, void *context) {
+    UA_Server *server = (UA_Server*)context;
+    UA_GDSTransaction *transaction = &server->transaction;
+    UA_Boolean closeRsaMin = false;
+    UA_Boolean closeRsaSha256 = false;
+
+    /* Check which secureChannels are affected and close them */
+    UA_NodeId rsaShaCertificateType = UA_NODEID_NUMERIC(0, UA_NS0ID_RSASHA256APPLICATIONCERTIFICATETYPE);
+    UA_NodeId rsaMinCertificateType = UA_NODEID_NUMERIC(0, UA_NS0ID_RSAMINAPPLICATIONCERTIFICATETYPE);
+
+    for(size_t i = 0; i < transaction->certificateInfosSize; i++) {
+        UA_GDSCertificateInfo certInfo = transaction->certificateInfos[i];
+        if(UA_NodeId_equal(&certInfo.certificateType, &rsaMinCertificateType))
+            closeRsaMin = true;
+        if(UA_NodeId_equal(&certInfo.certificateType, &rsaShaCertificateType))
+            closeRsaSha256 = true;
+    }
+
+    UA_NodeId defaultApplicationGroup =
+            UA_NODEID_NUMERIC(0, UA_NS0ID_SERVERCONFIGURATION_CERTIFICATEGROUPS_DEFAULTAPPLICATIONGROUP_TRUSTLIST);
+
+    for(size_t i = 0; i < transaction->certGroupSize; i++) {
+        UA_CertificateGroup certGroup = transaction->certGroups[i];
+        if(!UA_NodeId_equal(&certGroup.certificateGroupId, &defaultApplicationGroup))
+            continue;
+        closeRsaMin = true;
+        closeRsaSha256 = true;
     }
```

Comment:
This may be a little too offensive. When the trust list of the DefaultApplicationGroup has been modified, only the secure channels where the certificate of the remote peer is not trusted anymore shall be closed (see https://reference.opcfoundation.org/GDS/v105/docs/7.10.6). After adding certificates for example no secure channel needs to be touched.

There is also an additional issue:
While it is valid to leave any session untouched when only UpdateCertificate() has been called, a change to the TrustList of the DefaultUserTokenGroup must be accompanied by closing all Sessions with a now untrusted / revoked certificate (see https://reference.opcfoundation.org/GDS/v105/docs/7.8.2.3). As of Spec version 1.5.4 also the underyling SecureChannels have to be closed in this case.

Reply:
For this reason, the SecureChannel is simply closed if the server certificate or the TrustList changes. The client can re-establish the connection with its reconnect logic and its certificate is re-validated against the new TrustList.
This makes the whole thing relatively easy to implement.

Section of the CloseAndUpdate function:
"When the TrustList changes the Server shall re-evaluate the Certificate associated with any open Sessions. Sessions with an untrusted or revoked Certificate shall be closed."

I believe this only applies when no transaction logic is in place. With transaction logic, the changes are only applied during the apply process.

Reply:
Of course it is easy to implement. Nevertheless it leads to interruption where it is not required. 
Still certificate-authenticated sessions remain intact which in case of changes to the TrustList of the DefaultUserTokenGroup is not desired. Affected sessions shall be closed as well in this case.